### ISSUE TITLE
G2O segmentation fault with updated master branch code

### ISSUE THREAD

**Issue by suyash023** [RainerKuemmerle/g2o#610]
Hello! I am currently using G2O to perform bundle-adjustment. I compiled G2O from scratch using the latest code on master branch. This is the configuration of my system:

- OS: Ubuntu-18.04
- Eigen version: 3.4 (also made attempts with Eigen 3.3)
- Suitesparse downloaded using: sudo apt-get install libsuitesparse-dev
- g2o: latest code from master branch (added compilation flags: WITH_MARCH_NATIVE=ON and CMAKE_BUILD_TYPE=Release) (also made attempts with flags WITH_MARCH_NATIVE=OFF and CMAKE_BUILD_TYPE=Release)
- Application compiled with above flags as well, ensuring consistency in flags across application and g2o.

 The code compiles with no error, but I keep running into segmentation fault at different stages when iterating  with different solvers. This is the output of gdb: 
When running with solver CSparse linear solver:
[CSparse_solver.txt](https://github.com/RainerKuemmerle/g2o/files/9903024/CSparse_solver.txt)

Output of GDB when running with Cholmod:
[Cholmod_solver.txt](https://github.com/RainerKuemmerle/g2o/files/9903045/Cholmod_solver.txt)

Outptut of GDB when running with the PCG solver:
[PCG_solver.txt](https://github.com/RainerKuemmerle/g2o/files/9903058/PCG_solver.txt)

The fact that GDB output of different solvers output  is different and the crash happens at different stages, points to the idea that problem lies within the solvers. @RainerKuemmerle : Please advise further.

**Comment by RainerKuemmerle** [comment]
Are you sure that you are not running out of memory? Your stack traces are pointing to memory allocation issues?
Maybe the memory footprint of master increased compared to your former version.

**Comment by suyash023** [comment]
@RainerKuemmerle  I have 62.5 GB of memory and peak memory usage of the process in htop does not exceed 0.5 %. 
I dont think I am running out of memory.

**Comment by suyash023** [comment]
@RainerKuemmerle I have tried running my code with GaussNewton algorithm using core/optimization_algorithm_gauss_newton.h and it seems to work (but there is no convergence).

**Comment by suyash023** [comment]
@RainerKuemmerle Here is the function to clear the optimizer graph. I need the graph to be cleared after a certain number of measurements (local windowed bundle adjustment). Using the optimizer.clear() function also produces segmentation fault.
`void BundleAdjustment::resetBA() {
    //optimizer.clear(); //produces segmentation fault
    for (auto &edge : optimizer.edges()) {
            delete(edge);
        }
//    for (auto &vertex : optimizer.vertices()) {
//        delete (vertex.second);
//    }
    optimizer.edges().clear();
    optimizer.vertices().clear();
}`

**Comment by suyash023** [comment]
@RainerKuemmerle Here are the GDB logs when only optimizer.clear() is called.
[optimizer_clear.txt](https://github.com/RainerKuemmerle/g2o/files/9914297/optimizer_clear.txt)

**Comment by suyash023** [comment]
@RainerKuemmerle I have now modified my code design so that memory allocation does not happen in successive rounds of bundle adjustment. Information is just overwritten at the memory locations and optimizations are performed again. I had to set marginalization to false for land mark points in my case for it to work successful.

I had an additional question about camera parameters. Are distortion parameters also supported by g2o::CameraParameters ?
